rust/arrow/src/buffer.rs

```diff
@@ -50,15 +50,25 @@ struct BufferData {
     /// The raw pointer into the buffer bytes
     ptr: *const u8,
 
-    /// The length (num of bytes) of the buffer
+    /// The length (num of bytes) of the buffer. The region `[0, len)` of the buffer
+    /// is occupied with meaningful data, while the rest `[len, capacity)` is the
+    /// unoccupied region.
     len: usize,
+
+    /// The capacity (num of bytes) of the buffer
+    /// Invariant: len <= capacity
+    capacity: usize,
 }
 
 impl PartialEq for BufferData {
     fn eq(&self, other: &BufferData) -> bool {
         if self.len != other.len {
             return false;
         }
+        if self.capacity != other.capacity {
```

Comment:
I agree. It seems wrong that on one hand the memcmp is comparing up to `len` and we bypass that if there is a mismatch on `capacity`

Reply:
Thanks guys! I'm slightly confused - here we are comparing meaningful data, no? and if capacity mismatch then it is considered not equal and we skip comparing the data content.

The equality is defined as: 1) both length should be equal, 2) both capacity should be equal, and 3) data content up to length should be equal. Let me know if this definition sounds good to you.

Reply:
To me if two arrays only differ by the amount of padding that they have then I would consider them equal.  When I perform operations using these two arrays I will get the same answer (because the padding, or rather amount of padding, does not impact the result).  However:
 - I am focused on computation, maybe there are other implication in IPC, etc.
 - In practice, this probably won't come up as we round up padding to a multiple of 64 bytes, but it *could* happen.

Reply:
Array equality is defined separately in `array/equal.rs` and yes, it does take account on what you said (it compares buffer content with `data()`). In the current context we are discussing equality of buffers though, which IMO, when looking at in isolation, should consider both `len` and `capacity`.

Reply:
Ok, yes I agree.